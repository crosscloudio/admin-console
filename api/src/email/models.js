const {
  MAILJET_API_KEY,
  MAILJET_API_SECRET,
  MAILJET_SENDER_EMAIL,
  SMTP_SENDER_EMAIL,
  SMTP_URL,
} = process.env;

/**
 * Create a new email helper. It creates a mailjet version if its settings are
 * provided as environment variables or a debug one otherwise
 * @returns{BaseEmails}
 */
export function createEmailHelper() {
  let implementation;

  if (MAILJET_API_KEY && MAILJET_API_SECRET && MAILJET_SENDER_EMAIL) {
    implementation = require('./implementations/MailjetEmails'); // eslint-disable-line global-require
  } else if (SMTP_SENDER_EMAIL && SMTP_URL) {
    implementation = require('./implementations/SmtpEmails'); // eslint-disable-line global-require
  } else {
    implementation = require('./implementations/DebugEmails'); // eslint-disable-line global-require
  }

  // use `.default` because of es6 exports used in implementations
  return new implementation.default();
}
