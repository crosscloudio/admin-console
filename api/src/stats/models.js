import { keyBy, range, sumBy } from 'lodash';

/* eslint-disable camelcase */
// snake_case is used for compatibility with database

// mapping of storage identifiers to names to display in the admin console
const CSP_ID_NAME_MAPPING = {
  one: 'OneDrive',
  odb: 'onedrivebusiness',
  o365: 'office365groups',
  box: 'Box',
  dro: 'Dropbox',
  own: 'OwnCloud',
  gdr: 'Google Drive',
  nex: 'NextCloud',
  cif: 'Windows Share',
};

// Possible sync operation statuses
const SYNC_OPERATION_STATUSES = ['success', 'blocked', 'failed'];

// Sunday is 0 according to the result of Postgresql `extract(dow from column)`
// https://www.postgresql.org/docs/current/static/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT
const WEEKDAYS = [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
];

/**
 * A class which instances are created as resolvers for the `Stats` type.
 * All methods return promises which resolves to the actual stats data.
 */
export class StatsHelper {
  /**
   * Create a StatsHelper
   * @param {Object} knex - an instance of the knex.js object
   * @param {Object} -  an object with models / DAO objects returned
   * by the `createModels` function
   * @param {number|string} - the id of the current organization (used for
   * filtering)
   */
  constructor(knex, models, organization_id) {
    this.knex = knex;
    this.models = models;
    this.organization_id = organization_id;
  }

  async active_users_today() {
    const { knex } = this;
    // getting start date of today
    const startToday = new Date();
    startToday.setHours(0, 0, 0, 0);

    // getting stop date of today
    const stopToday = new Date();
    stopToday.setHours(23, 59, 59, 999);

    const result = await knex.raw(
      `
      SELECT COUNT(*) FROM (
        SELECT DISTINCT user_id FROM activity_logs WHERE
        organization_id = ? AND
        timestamp >= ? AND
        timestamp <= ?
      ) AS count;
    `,
      [this.organization_id, startToday, stopToday]
    );
    return result.rows[0].count;
  }

  async encryption_status() {
    const { knex } = this;
    const result = await this.getFilteredActivityLogs()
      .select('encrypted', knex.raw('count(*) AS count'))
      .groupBy('encrypted');
    const resultByEncryptionStatus = keyBy(result, 'encrypted');
    const getCount = encryped =>
      resultByEncryptionStatus[encryped]
        ? resultByEncryptionStatus[encryped].count
        : 0;
    return {
      encrypted_files: getCount(true),
      not_encrypted_files: getCount(false),
    };
  }

  file_type_usage() {
    const { knex } = this;
    return this.getFilteredActivityLogs()
      .select(knex.raw('mime_type AS type'), knex.raw('count(*) AS count'))
      .groupBy('mime_type')
      .orderBy('count', 'desc')
      .limit(10); // only 10 items are displayed in the dashboard
  }

  policy_usage() {
    const { knex } = this;
    return this.models.policies
      .getTable()
      .where({
        organization_id: this.organization_id,
      })
      .select('type', knex.raw('count(*) AS count'))
      .groupBy('type');
  }

  async sync_operations_status() {
    const { knex } = this;
    const result = await this.getFilteredActivityLogs()
      .select(knex.raw('status AS type'), knex.raw('count(*) AS count'))
      .groupBy('status');

    // statuses returned from database
    const returnedStatuses = new Set(result.map(stats => stats.type));

    // ensure all default operation statuses are in the returned data
    SYNC_OPERATION_STATUSES.forEach(status => {
      if (!returnedStatuses.has(status)) {
        result.push({
          type: status,
          count: 0,
        });
      }
    });
    return result;
  }

  traffic_stats() {
    const { knex } = this;
    return this.getFilteredActivityLogs()
      .select(
        'type',
        knex.raw('count(*) AS count'),
        knex.raw('sum(bytes_transferred) AS traffic')
      )
      .groupBy('type');
  }

  async usage_last_week() {
    const { knex } = this;

    // getting last 7 days usage
    const oneWeekAgo = new Date();
    oneWeekAgo.setDate(oneWeekAgo.getDate() - 6);
    oneWeekAgo.setHours(0, 0, 0, 0);

    const result = await this.getFilteredActivityLogs()
      .where('timestamp', '>=', oneWeekAgo)
      .select(
        knex.raw('extract(dow from timestamp) AS day_of_week'),
        knex.raw('count(*) AS count'),
        knex.raw('sum(bytes_transferred) AS traffic')
      )
      .groupByRaw('extract(dow from timestamp)');

    const resultByDateOfWeek = keyBy(result, 'day_of_week');
    // we need all days of week, even if there was no activity for that day
    const resultForAllWeekdays = range(7).map(day => {
      return resultByDateOfWeek[day] || { day_of_week: day };
    });

    // which day was one week ago
    const firstDay = oneWeekAgo.getDay();
    const resultWithCorrectOrder = [
      ...resultForAllWeekdays.slice(firstDay),
      ...resultForAllWeekdays.slice(0, firstDay),
    ];

    return resultWithCorrectOrder.map(usage => ({
      day: WEEKDAYS[usage.day_of_week],
      count: usage.count || 0,
      traffic: usage.traffic || 0,
    }));
  }

  async used_storages() {
    const { knex } = this;
    const result = await this.getFilteredActivityLogs()
      .select(
        // TODO: add a functional index or store csp_id in a separate column
        knex.raw('substr(storage_id, 0, 4) AS csp_id'),
        knex.raw('count(*) AS count')
      )
      .groupByRaw('substr(storage_id, 0, 4)');

    // summing up the logs to calculate percentages
    // warning: `count` is returned from the database as a string
    const totalTasks = sumBy(result, storage => parseFloat(storage.count));
    return result
      .filter(
        // only return data for csps with name mapping
        storage => CSP_ID_NAME_MAPPING[storage.csp_id]
      )
      .map(storage => ({
        name: CSP_ID_NAME_MAPPING[storage.csp_id],
        count: storage.count,
        // percentage of total usage rounded to two decimals
        percentage:
          Math.round(
            storage.count / totalTasks * 100 * 100 // eslint-disable-line no-mixed-operators
          ) / 100,
      }));
  }

  getFilteredActivityLogs() {
    return this.models.activityLogs.getTable().where({
      organization_id: this.organization_id,
    });
  }
}
