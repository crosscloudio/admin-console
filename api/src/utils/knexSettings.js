let PG_URL = process.env.DATABASE_URL || 'postgres://postgres:@db/postgres';

// required for heroku
if (process.env.DATABASE_FORCE_SSL) {
  PG_URL += '?ssl=true';
}

export default {
  client: 'pg',
  connection: PG_URL,
  migrations: {
    tableName: 'knex_migrations',
  },
};
