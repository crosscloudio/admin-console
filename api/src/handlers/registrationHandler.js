import Yup from 'yup';
import zxcvbn from 'zxcvbn';

import { ZXCVBN_USER_INPUTS } from '../utils/zxcvbnInputs';
import { loginUser } from '../utils/auth';
import rateLimit from '../utils/rateLimit';
import yupToFormErrors from '../utils/yupToFormErrors';

function createRegistrationSchema(models) {
  return Yup.object().shape({
    user_name: Yup.string().required('Please provide your name'),
    company_name: Yup.string().test(
      'check-company-name-availability',
      'Company with this name already exists',
      name => {
        if (!name) {
          return true;
        }
        return models.organizations.checkIfNameAvailable(name);
      }
    ),
    email: Yup.string()
      .required('Email is required')
      .email('Invalid email address')
      .test(
        'check-name-availability',
        'User with this email already exists',
        email => models.users.checkIfEmailAvailable(email)
      ),
    password: Yup.string()
      .required('Please provide password')
      .min(
        8,
        'Password should have at least ${min} chars' // eslint-disable-line no-template-curly-in-string
      )
      .test('password-zxcvbn', 'Password is not strong enough', function(
        password
      ) {
        // don't allow to provide `CrossCloud` or similar strings and also
        // the email address of the user
        const userInputs = [...ZXCVBN_USER_INPUTS];
        if (this.parent.email) {
          userInputs.push(this.parent.email);
        }
        const passwordCheckResult = zxcvbn(password, userInputs);
        return passwordCheckResult.score >= 3;
      }),
  });
}

export default async function registrationHandler(ctx) {
  if (typeof ctx.request.body !== 'object') {
    ctx.throw(400, 'Please provide registration data');
  }

  const schema = createRegistrationSchema(ctx.models);
  let data;
  try {
    data = await schema.validate(ctx.request.body, { abortEarly: false });
  } catch (error) {
    ensureValidationErrors(error);
    ctx.body = { errors: yupToFormErrors(error) };
    ctx.status = 422;
    return;
  }

  await rateLimit({
    category: 'register',
    id: ctx.ip,
    max: 10,
    duration: 3600000, // 1 hour
  });
  const newUser = await ctx.models.registrationHelper.register(data);
  const { token } = await loginUser({
    ip: ctx.ip,
    models: ctx.models,
    requireAdmin: true,
    user: newUser,
  });
  ctx.body = { token };
}

// Yup converts all errors to ValidationErrors if invoked with `abortEarly: false`
// hiding application errors. This function checks if it happened and throws
// previously hidden non validation errors
function ensureValidationErrors(baseError) {
  for (const error of baseError.inner) {
    for (const suberror of error.errors) {
      if (typeof suberror !== 'string' && suberror.name !== 'ValidationError') {
        throw suberror;
      }
    }
  }
}
