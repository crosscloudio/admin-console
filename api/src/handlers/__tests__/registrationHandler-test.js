import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import knex from 'knex';
import request from 'supertest-as-promised';

import { createModels } from '../../models';
import knexSettings from '../../utils/knexSettings';
import registrationHandler from '../registrationHandler';

describe('registrationHandler', () => {
  let app;
  let knexInstance;
  let transaction;
  let models;

  beforeEach(async () => {
    knexInstance = knex(knexSettings);
    await new Promise((resolve, reject) => {
      knexInstance
        .transaction(trx => {
          // WARNING: don't use async/await in this callback (or return a promise)
          // or the transaction will be automatically commited
          transaction = trx;
          models = createModels({ knex: transaction });
          resolve();
        })
        .catch(reject);
    });
    Date.now = jest.fn(
      // Tue May 30 2017 10:40:16 GMT+0200 (CEST)
      () => 1496133616539
    );

    app = new Koa();
    app.use(bodyParser());

    // make models available in the app context
    app.use((ctx, next) => {
      ctx.models = models;
      return next();
    });
    app.use(registrationHandler);
  });

  afterEach(async () => {
    await transaction.rollback();
    await knexInstance.destroy();
  });

  it('should throw if the data is incorrect', async () => {
    const response = await request(app.listen()).post('/').send({}).expect(422);
    expect(response.body).toMatchSnapshot();
  });

  it('should create a user and an organization if the data is incorrect', async () => {
    const response = await request(app.listen())
      .post('/')
      .send({
        email: 'user@example.com',
        password: 'a some very complicated password',
        user_name: 'John Smith',
      })
      .expect(200);
    expect(response.body).toBeTruthy();
    expect(response.body.token).toBeTruthy();
  });
});
