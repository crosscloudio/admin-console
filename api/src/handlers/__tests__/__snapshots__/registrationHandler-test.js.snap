// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registrationHandler should throw if the data is incorrect 1`] = `
Object {
  "errors": Object {
    "email": "Email is required",
    "password": "Please provide password",
    "user_name": "Please provide your name",
  },
}
`;
